### PR TITLE
Add Roberta SQuAD model

### PR DESCRIPTION
**Pull request by vlarine**
There is the realisation of a RoBERTa SQuAD finetuning.

On 2x1080Ti on RoBERTa Base it gives:

python3 run_squad.py \
  --model_type roberta \
  --model_name_or_path roberta-base \
  --do_train \
  --do_eval \
  --train_file $SQUAD_DIR/train-v1.1.json \
  --predict_file $SQUAD_DIR/dev-v1.1.json \
  --per_gpu_train_batch_size 8 \
  --per_gpu_eval_batch_size 8 \
  --learning_rate 3e-5 \
  --num_train_epochs 2.0 \
  --max_seq_length 384 \
  --doc_stride 128 \
  --save_steps 2000 \
  --overwrite_output_dir \
  --verbose_logging \
  --output_dir /tmp/debug_squad/
Results: {'exact': 85.80889309366131, 'f1': 92.09291402361669, 'total': 10570, 'HasAns_exact': 85.80889309366131, 'HasAns_f1': 92.09291402361669,   'HasAns_total': 10570}

On RoBERTa Large:

python3 run_squad.py \
  --model_type roberta \
  --model_name_or_path roberta-large \
  --do_train \
  --do_eval \
  --train_file $SQUAD_DIR/train-v1.1.json \
  --predict_file $SQUAD_DIR/dev-v1.1.json \
  --per_gpu_train_batch_size 2 \
  --per_gpu_eval_batch_size 2 \
  --learning_rate 3e-5 \
  --num_train_epochs 2.0 \
  --max_seq_length 384 \
  --doc_stride 128 \
  --save_steps 2000 \
  --overwrite_output_dir \
  --verbose_logging \
  --output_dir /tmp/debug_squad/
Results: {'exact': 87.04824976348155, 'f1': 93.14253401654709, 'total': 10570, 'HasAns_exact': 87.04824976348155, 'HasAns_f1': 93.14253401654709,   'HasAns_total': 10570}

### CODE REVIEW

File: examples/utils_squad.py [hunk 1]
```diff
@@ -908,9 +929,7 @@ def _strip_spaces(text):
     # and `pred_text`, and check if they are the same length. If they are
     # NOT the same length, the heuristic has failed. If they are the same
     # length, we assume the characters are one-to-one aligned.
-    tokenizer = BasicTokenizer(do_lower_case=do_lower_case)
-
-    tok_text = " ".join(tokenizer.tokenize(orig_text))
+    tok_text = " ".join(tokenizer.tokenize(orig_text)) if tokenizer else orig_text
```
Comment (ikuyamada) [inline examples/utils_squad.py]: Here, we need to use `BasicTokenizer` instead of model-specific tokenizer which can split a token into multiple tokens, and inject wordpiece prefixes (`##`).

**Comment by ikuyamada** [comment]
I am also working on reproducing the results reported in the Roberta paper and found two issues in this PR. One issue is explained in the comment above. The other issue is that it is required to insert two sep_tokens between question tokens and answer tokens for Roberta as implemented [here](https://github.com/huggingface/transformers/blob/master/transformers/tokenization_roberta.py#L101). Therefore, `max_tokens_for_doc` should be `max_seq_length - len(query_tokens) - 4`.

File: examples/utils_squad.py (hunk 1 above)
Comment (vlarine) [reply examples/utils_squad.py]: But `BasicTokenizer` has its own vocabulary. Isn't a conflict with a model-specific tokenizer?

**Comment by vlarine** [comment]
In the Fairseq realisation of RoBERTa on Commonsense QA:
https://github.com/pytorch/fairseq/tree/master/examples/roberta/commonsense_qa
https://github.com/pytorch/fairseq/blob/master/examples/roberta/commonsense_qa/commonsense_qa_task.py

There is the only one sep_token between question and answer:
`<s> Q: Where would I not want a fox? </s> A: hen house </s>`

File: examples/utils_squad.py (hunk 1 above)
Comment (ikuyamada) [reply examples/utils_squad.py]: If I understand correctly, `BasicTokenizer` does not have its own vocabulary. It is just a simple rule-based tokenizer with text normalization.

Comment (vlarine) [reply examples/utils_squad.py]: Yes, you are right. There we need to keep `BasicTokenizer`.

**Comment by ikuyamada** [comment]
> In the Fairseq realisation of RoBERTa on Commonsense QA:
> There is the only one sep_token between question and answer:
> `<s> Q: Where would I not want a fox? </s> A: hen house </s>`

Thank you very much for your prompt reply. I did not know this. It seems to be appropriate to use single `sep_token` here because Commonsense QA is somewhat more similar to SQuAD than other tasks (e.g., GLUE).

**Comment by julien-c** [comment]
Thanks for this @vlarine! (and @ikuyamada)

Would you agree to share the weights on our S3 as well?

Also, did you try with the same separators encoding scheme as the other RoBERTa models? 
`<s> Q: Where would I not want a fox? </s> </s> A: hen house </s>` – did the results differ significantly?

**Comment by vlarine** [comment]
No, I have not tried. But why there are two `</s>` tokens? I think more natural way is:
`<s> Q: Where would I not want a fox? </s> <s> A: hen house </s>`

**Comment by julien-c** [comment]
@vlarine See this docstring in `fairseq`: https://github.com/pytorch/fairseq/pull/969/files

Do you think you could try with this sep encoding scheme? Otherwise I'll do it in the next couple of days.

I would like to merge your PR soon. Any way you can give me write access to your fork, cf. https://help.github.com/en/articles/committing-changes-to-a-pull-request-branch-created-from-a-fork – so that i can add commits on top of your PR?

File: transformers/modeling_roberta.py [hunk 2]
```diff
@@ -43,9 +43,6 @@ class RobertaEmbeddings(BertEmbeddings):
     def __init__(self, config):
         super(RobertaEmbeddings, self).__init__(config)
         self.padding_idx = 1
-        self.word_embeddings = nn.Embedding(config.vocab_size, config.hidden_size, padding_idx=self.padding_idx)
-        self.position_embeddings = nn.Embedding(config.max_position_embeddings, config.hidden_size,
-                                                padding_idx=self.padding_idx)
```
Comment (julien-c) [inline transformers/modeling_roberta.py]: extraneous delete

File: transformers/modeling_roberta.py [hunk 3]
```diff
@@ -24,7 +24,7 @@
 import torch.nn as nn
 from torch.nn import CrossEntropyLoss, MSELoss
 
-from .modeling_bert import BertEmbeddings, BertLayerNorm, BertModel, BertPreTrainedModel, gelu
+from .modeling_bert import BertEmbeddings, BertLayerNorm, BertModel, BertPreTrainedModel, BertForQuestionAnswering, gelu
```
Comment (julien-c) [inline transformers/modeling_roberta.py]: unneeded import

File: transformers/__init__.py [hunk 4]
```diff
@@ -74,15 +74,15 @@
                                 GPT2LMHeadModel, GPT2DoubleHeadsModel,
                                 load_tf_weights_in_gpt2, GPT2_PRETRAINED_MODEL_ARCHIVE_MAP)
     from .modeling_xlnet import (XLNetPreTrainedModel, XLNetModel, XLNetLMHeadModel,
-                                XLNetForSequenceClassification, XLNetForMultipleChoice,
-                                XLNetForQuestionAnsweringSimple, XLNetForQuestionAnswering,
```
Comment (julien-c) [inline transformers/__init__.py]: extraneous deletes in this file

**Comment by pminervini** [comment]
@vlarine @julien-c thanks for the amazing work! I'll try it on SQuAD 2.0 and let you know

**Comment by erenup** [comment]
Nice work. I also tried to add roberta into run_squad.py several days ago. Hope that my implementation would be useful. [run_squad.py with roberta](https://github.com/erenup/pytorch-transformers/pull/4)

**Comment by julien-c** [comment]
Folding this PR into #1386, which is close to being ready to being merged.

@vlarine @ikuyamada @pminervini @erenup Can you guys please check it out?

File: transformers/__init__.py [hunk 5]
```diff
@@ -74,15 +74,15 @@
...
+                                XLNetForSequenceClassification, XLNetForQuestionAnsweringSimple,
+                                XLNetForQuestionAnswering,
                                 load_tf_weights_in_xlnet, XLNET_PRETRAINED_MODEL_ARCHIVE_MAP)
     from .modeling_xlm import (XLMPreTrainedModel , XLMModel,
                             XLMWithLMHeadModel, XLMForSequenceClassification,
                             XLMForQuestionAnswering, XLMForQuestionAnsweringSimple,
                             XLM_PRETRAINED_MODEL_ARCHIVE_MAP)
-    from .modeling_roberta import (RobertaForMaskedLM, RobertaModel,
-                                RobertaForSequenceClassification, RobertaForMultipleChoice,
```
Comment (caspillaga) [inline transformers/__init__.py]: Why did you remove RobertaForMultipleChoice?

**Comment by julien-c** [comment]
Closing in favor of #1386.
